src/ray/gcs/gcs_server/gcs_actor_manager.cc

```diff
@@ -422,12 +422,17 @@ Status GcsActorManager::RegisterActor(const ray::rpc::RegisterActorRequest &requ
   }
 
   actor_to_register_callbacks_[actor_id].emplace_back(std::move(success_callback));
-  RAY_CHECK(registered_actors_.emplace(actor->GetActorID(), actor).second);
+
+  // If GCS restarts before reply to GCS client and after writes to storage,
+  // GCS client will resend the `RegisterActor` request. After GCS restarts,
+  // `registered_actors_` and `unresolved_actors_` may recover data from storage, so we
+  // override them directly without checking.
+  registered_actors_.emplace(actor->GetActorID(), actor);
 
   const auto &owner_address = actor->GetOwnerAddress();
   auto node_id = ClientID::FromBinary(owner_address.raylet_id());
   auto worker_id = WorkerID::FromBinary(owner_address.worker_id());
-  RAY_CHECK(unresolved_actors_[node_id][worker_id].emplace(actor->GetActorID()).second);
+  unresolved_actors_[node_id][worker_id].emplace(actor->GetActorID());
```

Comment:
We remove CHECK because of the same reason as above?

Reply:
Yeah, I explained it in the comments on line 428.